go/tasks/plugins/webapi/bigquery/plugin.go

```diff
@@ -0,0 +1,512 @@
+	}
+
+	return &ResourceWrapper{
+		Status: job.Status,
+	}, nil
+}
+
+func (p Plugin) Delete(ctx context.Context, taskCtx webapi.DeleteContext) error {
+	if taskCtx.ResourceMeta() == nil {
+		return nil
+	}
+
+	resourceMeta := taskCtx.ResourceMeta().(*ResourceMetaWrapper)
+
+	identity := google.Identity{
+		K8sNamespace:      resourceMeta.Namespace,
+		K8sServiceAccount: resourceMeta.K8sServiceAccount,
+	}
+	client, err := p.newBigQueryClient(ctx, identity)
+
+	if err != nil {
+		return err
+	}
+
+	_, err = client.Jobs.Cancel(resourceMeta.JobReference.ProjectId, resourceMeta.JobReference.JobId).Do()
+
+	if err != nil {
+		return err
+	}
+
+	logger.Info(ctx, "Cancelled job [%s]", formatJobReference(resourceMeta.JobReference))
+
+	return nil
+}
+
+func (p Plugin) Status(_ context.Context, tCtx webapi.StatusContext) (phase core.PhaseInfo, err error) {
+	resourceMeta := tCtx.ResourceMeta().(*ResourceMetaWrapper)
+	resource := tCtx.Resource().(*ResourceWrapper)
+	version := pluginsCore.DefaultPhaseVersion
+
+	if resource == nil {
+		return core.PhaseInfoUndefined, nil
+	}
+
+	taskInfo := createTaskInfo(resourceMeta)
+
+	if resource.CreateError != nil {
+		return handleCreateError(resource.CreateError, taskInfo), nil
+	}
+
+	switch resource.Status.State {
+	case "PENDING":
+		return core.PhaseInfoQueuedWithTaskInfo(version, "Query is PENDING", taskInfo), nil
+
+	case "RUNNING":
+		return core.PhaseInfoRunning(version, taskInfo), nil
+
+	case "DONE":
+		if resource.Status.ErrorResult != nil {
+			return handleErrorResult(
+				resource.Status.ErrorResult.Reason,
+				resource.Status.ErrorResult.Message,
+				taskInfo), nil
+		}
+
+		return pluginsCore.PhaseInfoSuccess(taskInfo), nil
+	}
+
+	return core.PhaseInfoUndefined, pluginErrors.Errorf(pluginsCore.SystemErrorCode, "unknown execution phase [%v].", resource.Status.State)
+}
+
+func handleCreateError(createError *googleapi.Error, taskInfo *core.TaskInfo) core.PhaseInfo {
+	code := fmt.Sprintf("http%d", createError.Code)
+
+	userExecutionError := &flyteIdlCore.ExecutionError{
+		Message: createError.Message,
+		Kind:    flyteIdlCore.ExecutionError_USER,
+		Code:    code,
+	}
+
+	systemExecutionError := &flyteIdlCore.ExecutionError{
+		Message: createError.Message,
+		Kind:    flyteIdlCore.ExecutionError_SYSTEM,
+		Code:    code,
+	}
+
+	if createError.Code >= http.StatusBadRequest && createError.Code < http.StatusInternalServerError {
+		return core.PhaseInfoFailed(pluginsCore.PhasePermanentFailure, userExecutionError, taskInfo)
+	}
+
+	if createError.Code >= http.StatusInternalServerError {
+		return core.PhaseInfoFailed(pluginsCore.PhaseRetryableFailure, systemExecutionError, taskInfo)
+	}
+
+	// something unexpected happened, just terminate task
+	return core.PhaseInfoFailed(pluginsCore.PhasePermanentFailure, systemExecutionError, taskInfo)
+}
+
+func handleErrorResult(reason string, message string, taskInfo *core.TaskInfo) core.PhaseInfo {
+	userExecutionError := &flyteIdlCore.ExecutionError{
```

Comment:
also we have helper methods for system and user errors in the same file
https://github.com/flyteorg/flyteplugins/blob/e7bf9a4e1b43ec7be55411fcc31babbe6c769db4/go/tasks/pluginmachinery/core/phase.go#L188

Reply:
Oh, nice! Thanks